### ISSUE TITLE
Critically bad performance during ticket sale

### ISSUE THREAD

**Issue by ripper234** [Midburn/burning#45]
I am proposing this issue becomes the focal point for public discussion of this known & painful state of the ticketing system.

Fact: the ticket sale system barely functions. It does fulfill its purpose, but at a huge cost of people's time, frustration, and energy.

Can anyone shed some light on what exactly has been done to profile the ticket sale procedure? Where is the biggest performance bottleneck?

----

Requirements from the ticket sale system
----------------------------------------------------

1. **Fairness** - the ticket should not give each burner an equal shot at a ticket
2. **Resistance to hacking** - it should be very difficult to cheat the system
3. **Load** - the system needs to handle around 10,000 users a.t.m (with an upgrade path to reach 20,000 within 5 years).
4. **Latency** - I will use an unusual definition of latency here: the system should not waste people's time. In other words, someone should be able to participate in the ticket lottery with about a total time investment of up to 5-10 minutes (preparing the profile, making the credit card payment, etc.). Today people spend closer to 1-2 hours prepping, refreshing the page, complaining about it on Facebook, etc.
5. **Auditability** - We should be able to prove (in the Midburn transparency report) that the ticket sale was fair.

**Comment by ripper234** [comment]
I'd like to propose a quick & "easy" fix, and have someone debunk it. Odds are this doesn't work, but I don't see the exact problem with my proposal:

1. Use a secure, auditable randomness source. One such source is the Bitcoin Blockchain (the hash of the blocks is cryptographically random and too computationally expensive to predict/forge). To KISS we can just use a non-auditable randomness source to begin with e.g. [java.security.SecureRandom](https://docs.oracle.com/javase/7/docs/api/java/security/SecureRandom.html).
2. Make sure there are no Fake profiles (captchas, [The Burner Games](https://github.com/Midburn/burner-games), pre-added credit cards are some ways to implement this).
3. Shuffle all Midburn profiles using a random seed from the randomness generator using something like the [Knuth Shuffle](https://en.wikipedia.org/wiki/Random_permutation#Knuth_shuffles).

Does this solve the problem? Or am I missing something?

**Comment by Asisalhov** [comment]
In my opinion, the security emphasis is a bit of an overkill, like building concrete walls along the entrance queue in the event.
I think that sinplicity and agility are the main points that needs to be addressed.

**Comment by ripper234** [comment]
Good point @Asisalhov , I added Simplicity as the second requirement. What I described clearly takes some effort (another dedicated hackathon for this, perhaps?). Who here has reviewed the code we use today, and can say what is a good solution that answers 1 & 2?

**Comment by ripper234** [comment]
However I do want to state that I believe my proposed solution really isn't *that complicated*.

If Midburn IT is on board, what I would do is proceed to write a design/architecture document of the solution, and then do a dedicated hackathon just for this. Everybody coming in to this hackathon would already have the required skills, because we pre-select and pre-advertise what specific skills we need to implement this.

For a skilled team and with a bit of planning, it's possible to deploy a prototype of this within a 48 hour hackathon, including some basic load testing scripts in place. The solution won't be production ready within that time frame, of course, but we'll have a start.

**Comment by Asisalhov** [comment]
Here is an excellent article about high availability:
http://highscalability.com/blog/2013/12/16/22-recommendations-for-building-effective-high-traffic-web-s.html

**Comment by RoySegall** [comment]
I afraid this loop won't affect the near sale but as i mentioned before - the very far future. The system does not have any relate to the failing of the sail but as I can see it the server.

I believe that amazon cloud system could not handle Drupal hosting as much as pantheon, platform.sh or even acquia. This are the type of company that know how to handle drupal hosting at the best.

**Comment by ripper234** [comment]
FYI we are working on another, much much simpler solution that can be integrated in time for the near ticket sale next month. Details TBD.
